addons/link_tracker/models/mail_render_mixin.py

```diff
@@ -34,26 +35,28 @@ def _shorten_links(self, html, link_tracker_vals, blacklist=None, base_url=None)
 
         :return: updated html
         """
+        if not html or is_html_empty(html):
+            return html
         base_url = base_url or self.env['ir.config_parameter'].sudo().get_param('web.base.url')
         short_schema = base_url + '/r/'
-        for match in set(re.findall(tools.HTML_TAG_URL_REGEX, html)):
-            long_url = match[1]
-            # Don't shorten already-shortened links
-            if long_url.startswith(short_schema):
-                continue
+
+        def should_convert_to_tracker(url):
+            if re.search(rf'^{URL_SKIP_PROTOCOL_REGEX}', url):
+                return False
+            if url.startswith(short_schema):
+                return False
             # Don't shorten urls present in blacklist (aka to skip list)
-            if blacklist and any(s in long_url for s in blacklist):
-                continue
-            label = (match[3] or '').strip()
+            if blacklist and any(s in url for s in blacklist):
+                return False
+            return True
 
-            create_vals = dict(link_tracker_vals, url=unescape(long_url), label=unescape(label))
+        def get_replacement_url(url, label):
```

Comment:
Don't like that, especially you give an inner method called "get_replacement_url" that actually creates link trackers ... why adding complexity layers ?

Reply:
This was expressly code that the caller passes, so they know what happens, but I agree that it requires readers to check another file if curious about the implementation.